2_download/src/fetch_wqp_data.R

```diff
@@ -1,195 +1,14 @@
-        mutate(task_num = cur_group_id()) %>% 
-        ungroup() %>%
-        mutate(pull_by_id = TRUE)
-    }) 
-  
-  # Assign a separate task number and download group for each site with bad ids
-  sitecounts_grouped_bad_ids <- sitecounts_bad_ids %>%
-    group_by(site_id) %>%
-    mutate(task_num = max(sitecounts_grouped_good_ids$task_num) + cur_group_id(),
-           pull_by_id = FALSE) %>%
-    ungroup()
-  
-  # Combine all sites back together (now with assigned download_grp id's) and
-  # format columns
-  sitecounts_grouped_out <- sitecounts_grouped_good_ids %>%
-    bind_rows(sitecounts_grouped_bad_ids) %>%
-    # Ensure the groups are ordered correctly by prepending a dynamic number of 0s
-    # before the task number based on the maximum number of tasks.
-    mutate(download_grp = sprintf(paste0("%s_%0", nchar(max(task_num)), "d"), 
-                                  grid_id, task_num)) %>% 
-    arrange(download_grp) %>%
-    select(site_id, lat, lon, datum, grid_id, CharacteristicName, results_count, 
-           download_grp, pull_by_id) 
-  
-  return(sitecounts_grouped_out)
-
-}
-
-
-
-#' @title Return a small bounding box around site(s)
-#' 
-#' @description 
-#' Some site identifiers contain undesired characters and cannot be
-#' parsed by WQP. This function creates a small bounding box around the 
-#' site(s) for the purposes of querying WQP by bBox rather than site id. 
-#' 
-#' @param sites data frame containing at minimum columns lat, lon, and datum.
-#' @param buffer_dist_degrees double; value indicating how large of a buffer
-#' should be added to the site(s). Bounding box will be computed in WGS84 and
-#' units are in degrees. Defaults to 0.005 degrees (~500 m).
-#' 
-#' @returns
-#' Returns an object of class "bbox" that represents the bounding box around
-#' the point locations included in `sites`. Contains vectors "xmin", "ymin",
-#' "xmax", and "ymax". 
-#' 
-#' @examples 
-#' sites <- data.frame(lon = c(-74.62238, -74.94351, -75.25741), 
-#'                     lat = c(39.79456, 39.31011, 39.52289),
-#'                     datum = rep('NAD83', 3))
-#' create_site_bbox(sites)
-#' 
-create_site_bbox <- function(sites, buffer_dist_degrees = 0.005){
-  
-  # assume unknown crs (datum equals "UNKNWN", "OTHER") correspond with WGS84
-  epsg_in <- case_when(unique(sites$datum) == "NAD83" ~ 4269,
-                       unique(sites$datum) == "WGS84" ~ 4326,
-                       unique(sites$datum) == "NAD27" ~ 4267,
-                       unique(sites$datum) == "UNKWN" ~ 4326,
-                       unique(sites$datum) == "OTHER" ~ 4326)
-  
-  # create a small buffer around the site(s) and then compute the bounding box
-  site_bbox <- sites %>%
-    sf::st_as_sf(coords = c("lon","lat"), crs = epsg_in) %>%
-    sf::st_transform(4326) %>%
-    sf::st_buffer(buffer_dist_degrees) %>%
-    sf::st_bbox()
-  
-  return(site_bbox)
-  
-}
-  
-
+source("2_download/src/retry.R")
```

Comment:
One pipeline-style question for you, @lindsayplatt. Regardless of which `retry_with_timeout` method I go with, I expect to have a separate R file that contains the retry functions. I've sourced that file at the top of  `2_download/src/fetch_wqp_data.R` so that if a user is inspecting `fetch_wqp_data()`, it will hopefully be more apparent where the retry handling steps are coming from. Does that make sense to you, or do you think it's more intuitive to include with the others in `2_download.R`?

Reply:
I like the separate retry script and I think it will 1) make it clear where those are coming from to users of this repo, and 2) make it easier for other folks to reuse this approach for their own download pipelines that aren't necessarily from the WQP.